File: sprint7/By languages/kotlin/lesson9code1.kt
// Заполняем массив нулями, так как длина НОВП в худшем случае равна нулю.
// Большую оценку без просмотра последовательностей мы не можем получить.
val dp = Array(n + 1) { IntArray(m + 1) }
for (i in 1..n) {
    for (j in 1..m) {
        if (A[i - 1] == B[j - 1]) {
            var maxLen = 0
            for (iPrev in 1 until i) {
                if (A[iPrev - 1] >= A[i - 1]) {
                    continue
                }
                for (jPrev in 1 until j) {
                    maxLen = maxOf(maxLen, dp[iPrev][jPrev])
                }
            }
            dp[i][j] = maxLen + 1
        }
    }
}

// Надо перебрать все пары концов НОВП, чтобы найти оптимальный ответ.
var answer = 0
for (i in 0..n) {
    for (j in 0..m) {
        answer = maxOf(answer, dp[i][j])
    }
}

File: sprint7/By languages/swift/lesson9code1.swift
var dp: [[Int]] = Array(repeating: Array(repeating: 0, count: m + 1), count: n + 1)

for i in 1...n {
    for j in 1...m {
        if A[i - 1] == B[j - 1] {
            var max_len = 0
            for i_prev in 1..<i {
                if A[i_prev - 1] >= A[i - 1] {
                    continue
                }
                for j_prev in 1..<j {
                    max_len = max(max_len, dp[i_prev][j_prev])
                }
            }
            dp[i][j] = max_len + 1
        }
    }
}

var answer = 0
for i in 0...n {
    for j in 0...m {
        answer = max(answer, dp[i][j])
    }
}

print(answer)
